src/container/srv_target.c

```diff
@@ -493,18 +450,126 @@ cont_start_agg_ult(struct ds_cont_child *cont)
 }
 
 static void
-cont_stop_agg_ult(struct ds_cont_child *cont)
+cont_stop_agg_ult(struct ds_cont_child *cont, struct sched_request *req)
 {
-	if (cont->sc_agg_req == NULL)
+	if (req == NULL)
 		return;
 
 	D_DEBUG(DB_EPC, DF_CONT"[%d]: Stopping aggregation ULT\n",
 		DP_CONT(cont->sc_pool->spc_uuid, cont->sc_uuid),
 		dss_get_module_info()->dmi_tgt_id);
 
-	sched_req_wait(cont->sc_agg_req, true);
-	sched_req_put(cont->sc_agg_req);
-	cont->sc_agg_req = NULL;
+	sched_req_wait(req, true);
+	sched_req_put(req);
+}
+
+static int
+cont_vos_aggregate_cb(struct ds_cont_child *cont, daos_epoch_range_t *epr,
+		      bool full_scan, struct agg_param *param)
+{
+	int rc;
+
+	rc = vos_aggregate(cont->sc_hdl, epr, ds_csum_recalc,
+			   agg_rate_ctl, param, full_scan);
+
+	/* Suppress csum error and continue on other epoch ranges */
+	if (rc == -DER_CSUM)
+		rc = 0;
+
+	/* Wake up GC ULT */
+	sched_req_wakeup(cont->sc_pool->spc_gc_req);
+	return rc;
+}
+
+static uint64_t
+cont_agg_start_eph_get(struct ds_cont_child *cont)
+{
+	vos_cont_info_t	cinfo;
+	int		rc;
+
+	/*
+	 * Query the 'Highest Aggregated Epoch', the HAE will be bumped
+	 * in vos_aggregate()
+	 */
+	rc = vos_cont_query(cont->sc_hdl, &cinfo);
+	if (rc) {
+		D_ERROR("cont query failed: rc: %d\n", rc);
+		return 0;
+	}
+
+	return cinfo.ci_hae;
+}
+
+static uint64_t
+cont_agg_max_epoch_get(struct ds_cont_child *cont)
+{
+	if (unlikely(ec_agg_disabled))
+		return DAOS_EPOCH_MAX;
```

Comment:
this is for VOS aggregation. not EC aggregation. If ec_agg is disabled, then it will return MAX,  i.e. VOS aggregation will not be bounded by EC aggregation boundary anymore.

Reply:
What is the purpose of introduce "ec_agg_disabled" flag?
for EC obj, we cannot do vos aggregation without finish EC aggregate ahead, that possibly cause data corruption, right?
For example, a full-stripe update with partial overwrite, if don't do EC agg ahead, and directly do VOS agg, it will mege the partial ext with full-stripe ext, without updating parity ext. That is incorrect behavior, no?

Reply:
oh, "ec_agg_disabled" is most likely for testing purpose or if the user does not want to use EC at all,  then the flag can help to disable these EC related ULT,  and remove those  overhead.
Yes, ec_agg_disabled can not be set after 2.0, when EC is fully functional.